arquillian-chameleon-runner/runner/src/main/java/org/arquillian/container/chameleon/runner/ArquillianChameleon.java

```diff
@@ -0,0 +1,74 @@
+package org.arquillian.container.chameleon.runner;
+
+import java.io.IOException;
+import java.lang.reflect.Method;
+import java.net.URL;
+import java.net.URLClassLoader;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.logging.Logger;
+import org.arquillian.container.chameleon.runner.extension.ChameleonRunnerAppender;
+import org.jboss.arquillian.junit.Arquillian;
+import org.junit.runner.notification.RunNotifier;
+import org.junit.runners.model.InitializationError;
+
+public class ArquillianChameleon extends Arquillian {
+
+    private static final Logger log = Logger.getLogger(ArquillianChameleon.class.getName());
+
+    public ArquillianChameleon(Class<?> testClass) throws InitializationError {
+        super(testClass);
+    }
+
+    @Override
+    public void run(RunNotifier notifier) {
+
+        Class<?> testClass = getTestClass().getJavaClass();
+
+        final ClassLoader parent = Thread.currentThread().getContextClassLoader();
+
+        if (isInClientSide(parent) && !isSpecialChameleonFile(parent)) {
+
+            try {
+                Path arquillianChameleonConfiguration = new ArquillianChameleonConfigurator().setup(testClass, parent);
+
+                log.info(String.format("Arquillian Configuration created by Chameleon runner is placed at %s.", arquillianChameleonConfiguration.toFile().getAbsolutePath()));
+
+                createSpecialChameleonFile(arquillianChameleonConfiguration.getParent());
+                addsArquillianFile(arquillianChameleonConfiguration.getParent(),
+                    parent);
+            } catch (Exception e) {
+                throw new IllegalArgumentException(e);
+            }
+        }
+        super.run(notifier);
+    }
+
+    /**
+     * We need to create an special file and put it inside classloader. This is because Chameleon runner adds configuration files dynamically inside the classpath.
+     * When you run your tests from your build tool, some or all of them shares the same classloader. So we need to avoid calling the same logic all the time to not get multiple files placed at classloader
+     * representing different versions, because then you have uncertainty on which configuration file is really used
+     * @param parent
+     * @throws IOException
+     */
+    private void createSpecialChameleonFile(Path parent) throws IOException {
```

Comment:
Call it `createChameleonMarkerFile`. Just wondering - is it thread safe? Should we be concerned about it at this point?

Reply:
I am not sure, but I think that as first approach adding a sync block would be better, it is not a long method so the block will be minimal.